### PR TITLE
inline documentation for working with the codec

### PR DESCRIPTION
**Pull request by EasterTheBunny**
Basic inline documentation describing how codecs are built and how they are expected to be used.

### CODE REVIEW

File: pkg/types/codec.go [hunk 1]
```diff
@@ -1,3 +1,63 @@
+/*
+Codec is an interface that provides encoding and decoding functionality for a specific type identified by a name.
+Because there are many types that a ContractReader or ContractWriter can either accept or return, all encoding
+instructions provided by the codec are based on the type name.
+
+Starting from the lowest level, take for instance a big.Int encoder where we want the output to be big endian binary
+encoded.
+
+	typeCodec, _ := binary.NewBigInt(32, true, binary.BigEndian())
```
Comment (winder) [inline pkg/types/codec.go:9]: I couldn't find `binary.NewBigInt`. It looks like this is supposed to be a simple example, but you don't show how to implement it. The godoc has a more complete `example_test.go`, you could reference this as `example_minimal_test.go`.

File: pkg/types/example_test.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	"fmt"
+	"math/big"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/codec/encodings"
+	"github.com/smartcontractkit/chainlink-common/pkg/codec/encodings/binary"
+)
+
+// Example provides a minimal example of constructing and using a codec. Refer to the codec docs in pkg/types/codec.go
+// for more explanation. For even more detail, refer to the Example in pkg/codec/example_test.go.
+func Example() {
```
[the remark this answers is not in the document]
Comment (EasterTheBunny) [reply pkg/types/example_test.go:14]: I tried this but I got this warning: `ExampleCodec refers to unknown identifier: Codec`. It seems like the example cannot be placed next to an interface?

[the remark this answers is not in the document]
Comment (jmank88) [reply pkg/types/example_test.go:14]: Warning from where?
What about `Example_Codec`?

[the remark this answers is not in the document]
Comment (EasterTheBunny) [reply pkg/types/example_test.go]: Well it complained but it worked. ¯\_(ツ)_/¯

File: pkg/types/codec.go [hunk 3]
```diff
@@ -29,6 +29,65 @@ type Decoder interface {
...
+GRPC interface where data may be JSON encoded, passed through GRPC, and JSON decoded on the other side.
+
+	decodedStruct, _ := codec.CreateType("SomeStruct", false)
+	_ = codec.Decode(encodedStructBytes, &decodedStruct, "SomeStruct")
+
+The `encodings` package provides a `Builder` interface that allows for the creation of any encoding type. This is useful
+for creating custom encodings such as the EVM ABI encoding. An encoder implements the `Builder` interface and plugs
+directly into `TypeCodec`.
+
+From the perspective of a `ContractReader` instance, the `itemType` at the top level is the `readIdentifier` which
```
Comment (jmank88) [inline pkg/types/codec.go:86]: Does this link properly? If not, you can try square brackets:
```suggestion
From the perspective of a [ContractReader] instance, the `itemType` at the top level is the `readIdentifier` which
```

File: pkg/types/codec.go [hunk 4]
```diff
@@ -29,6 +29,65 @@ type Decoder interface {
...
+This allows us to encode and decode big.Int values with big endian encoding using the `encodings.TypeCodec` interface.
+
+	encodedBytes := []byte{}
+
+	originalValue := big.NewInt(42)
+	encodedBytes, _ = typeCodec.Encode(originalValue, encodedBytes) // new encoded bytes are appended to existing
+
+	value, _, _ := typeCodec.Decode(encodedBytes, value)
+
+The additional encodings.TypeCodec methods such as `GetType() reflect.Type` allow composition. This is useful for
```
Comment (jmank88) [inline pkg/types/codec.go:51]: Might need square brackets on types refs to other packages?
```suggestion
The additional [encodings.TypeCodec] methods such as `GetType() reflect.Type` allow composition. This is useful for
```

File: pkg/types/example_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (winder) [reply pkg/types/example_test.go]: I wrote (and never published) a blog post on this subject.

> Examples are a type of test, so the file they live in must end with `_test.go`. Like tests, the simplest example is created with a single public function that follows a naming convention, specifically:
>   >  ExampleFunctionName_Suffix
>
> The **Example** function will be detected as part of the package documentation and placed in the **FunctionName** section of the documentation. If the **FunctionName** is omitted, the example will be displayed more prominantly, as a package-level example. Multiple examples may be provided for the package or each function by using a different **_Suffix** for each.

Ironically, the godoc documentation is pretty bad. The blog post from 2011 links to some godoc which links back to the blog post. [Looks like the Example syntax is documented in the testing package for some reason](https://pkg.go.dev/testing#hdr-Examples).

File: pkg/types/example_codec_test.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/codec/encodings"
+	"github.com/smartcontractkit/chainlink-common/pkg/codec/encodings/binary"
+)
+
+// ExampleCodec provides a minimal example of constructing and using a codec. For more detail, refer to the Example in
+// pkg/codec/example_test.go.
```
Comment (winder) [inline pkg/types/example_codec_test.go:13]: If we're going to commit to beefing up the godoc (yes please!), I don't think you need the extra reference here. Once we start answering questions by linking to the docs people will easily discover the other examples from the rendered docs.
```suggestion
// ExampleCodec provides a minimal example of constructing and using a codec.
```

**Review by Farber98 — APPROVED** [review]
This is good to have since the codec is somewhat complex. I find the example useful to complement the text

**Review by ilija42 — APPROVED** [review]

**Review by vyzaldysanchez — APPROVED** [review]

**Pull request merged**